[builder] Support for --skip-new-go-module (#10098)
open-telemetry/opentelemetry-collector: cmd/builder/README.md

```
to only execute the compilation step.

### Avoiding the use of a new go.mod file

> There is an additional option that controls one aspect of the build
> process, which specifically allows skipping the use of a new `go.mod`
> file.  When the `--skip-new-go-module` command-line flag is supplied,

Reviewer: I would keep it a bit simpler
```suggestion
You can optionally skip creating a new `go.mod` file.
When the `--skip-new-go-module` command-line flag is supplied,
```

It would also be nice to motivate this: when would an end-user be interested in using `--skip-new-go-module`?

the build process issues a `go get` command for each component,
relying on the Go toolchain to update the enclosing Go module
appropriately.

This command will avoid downgrading a dependency in the enclosing
module, according to
[`semver.Compare()`](https://pkg.go.dev/golang.org/x/mod/semver#Compare),
and will instead issue a log indicating that the component was not
upgraded.

> This mode cannot be used in conjunction with several features that
> control the generated `go.mod` file, including `replaces`, `excludes`,
> and the component `path` override.  For each of these features, users
> are expected to modify the enclosing `go.mod` directly.

Reviewer: nit: make it a bit more explicit

```suggestion
`--skip-new-go-module` is incompatible with `replaces`, `excludes`,
and the component `path` override.  For each of these features, users
are expected to modify the enclosing `go.mod` directly.
```
